media/server/gstplayer/source/tasks/generic/SetVolume.cpp

```diff
@@ -35,15 +43,74 @@ SetVolume::~SetVolume()
     RIALTO_SERVER_LOG_DEBUG("SetVolume finished");
 }
 
+firebolt::rialto::wrappers::rgu_Ease convertEaseTypeToRguEase(EaseType easeType)
+{
+    switch (easeType)
+    {
+    case EaseType::EASE_LINEAR:
+        return firebolt::rialto::wrappers::rgu_Ease::EaseLinear;
+    case EaseType::EASE_IN_CUBIC:
+        return firebolt::rialto::wrappers::rgu_Ease::EaseInCubic;
+    case EaseType::EASE_OUT_CUBIC:
+        return firebolt::rialto::wrappers::rgu_Ease::EaseOutCubic;
+    default:
+        throw std::invalid_argument("Unknown EaseType");
+    }
+
+    return firebolt::rialto::wrappers::rgu_Ease::EaseLinear;
+}
+
 void SetVolume::execute() const
 {
     RIALTO_SERVER_LOG_DEBUG("Executing SetVolume");
+
     if (!m_context.pipeline)
     {
         RIALTO_SERVER_LOG_ERROR("Setting volume failed. Pipeline is NULL");
         return;
     }
-    m_gstWrapper->gstStreamVolumeSetVolume(GST_STREAM_VOLUME(m_context.pipeline), GST_STREAM_VOLUME_FORMAT_LINEAR,
-                                           m_volume);
+    GstElement *audioSink = m_player.getSink(firebolt::rialto::MediaSourceType::AUDIO);
+
+    if (audioSink && m_glibWrapper->gObjectClassFindProperty(G_OBJECT_GET_CLASS(audioSink), "audio-fade"))
```

Comment:
Can you explain how normal set volume request is going to work? I suppose it will try to use this task, right? What will happen if the client will request normal change of volume from let's say 0.5 to 1.0? Is it going to use fade-volume property too?

Reply:
Yes so a normal set volume request will use the SetVolume task. So when the vol would change from 0.5 to 1.0, it would trigger the SetVol task and the constructor would take the target volume to be 1.0 (in this case), as well as the volDuration and easingType. So then if the sink exists and supports the audio-fade prop, then it will be used for transition, from 0.5 to 1.0. If the audio-fade prop is not available, it will check if the SoC supports audio-fade and apply the easing using doAudioEasingonSoc(). If neither are supported then it will use gstStreamVolumeSetVolume() without any audio fade effect. So looking at the fade-volume property in the netflix code, it is used in getFadeVolume, and in our case it would get the current volume, so yeh I think it would use the fade-volume property, but I can double check on that

Reply:
And why we want to use audio fade for the normal volume setting? Normal volume setting should apply the change immediately, not in gradual way as "audio-fade" does.

Reply:
Besides, is incremental audio fade actually a thing?

Reply:
Oh ok so we basically would want to skip the use of audio-fade in a normal volume change. So I can adjust the logic to only apply audio-fade when a fade-duration is explicitly requested, by including a flag to check that if the m_volumeDuration is 0, then it does an immediate change, otherwise it does volume transition using audio fade